etc/systemd/system-generators/zfs-mount-generator.in

```diff
@@ -73,31 +99,74 @@ process_line() {
   p_nbmand="${10}"
   p_encroot="${11}"
   p_keyloc="${12}"
+  p_systemd_noauto="${13}"
+  p_systemd_requires="${14}"
+  p_systemd_requiresmountsfor="${15}"
+  p_systemd_before="${16}"
+  p_systemd_after="${17}"
+  p_systemd_wantedby="${18}"
+  p_systemd_requiredby="${19}"
+  p_systemd_nofail="${20}"
+  p_systemd_ignore="${21}"
 
   # Minimal pre-requisites to mount a ZFS dataset
+  # By ordering before zfs-mount.service, we avoid race conditions.
+  after="zfs-import.target"
+  before="zfs-mount.service"
   wants="zfs-import.target"
+  requires=""
+  requiredmounts=""
+  wantedby=""
+  requiredby=""
+
+
+  if [ -n "${p_systemd_after}" ] && \
+      [ "${p_systemd_after}" != "-" ] ; then
+    after="${p_systemd_after} ${after}"
```

Comment:
When people upgrade, their `zfs-list.cache/*` files may lack `after`, `before`, `requires`, `requires-mounts-for`, `wantedby`, `requiredby`, etc. Please test for zero length strings here as well, so that upgrades don't result in broken configurations.

Reply:
> Please test for zero length strings here as well, so that upgrades don't result in broken configurations.

Uhm... what about the `-n` in line 123?

Reply:
Whoops, yeah. It was too early in the morning for me to be reviewing code, apparently.